### PR TITLE
[SPARK-9231] [MLlib] DistributedLDAModel method for top topics per document

### PR DESCRIPTION
**Pull request by hhbyyh**
jira: https://issues.apache.org/jira/browse/SPARK-9231

Helper method in DistributedLDAModel of this form:

```
/**
 * For each document, return the top k weighted topics for that document.
 * @return RDD of (doc ID, topic indices, topic weights)
 */
def topTopicsPerDocument(k: Int): RDD[(Long, Array[Int], Array[Double])]
```

### CODE REVIEW

**Comment by jkbradley** [comment]
@hhbyyh @rotationsymmetry  Could you please help review each other's patches for LDA?  I'm trying to maximize what we can merge by tomorrow!  : )  I'll try to take a look later today.  I appreciate it!
* [https://github.com/apache/spark/pull/7769]
* [https://github.com/apache/spark/pull/7785]

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala [hunk 1]
```diff
@@ -561,6 +561,20 @@ class DistributedLDAModel private[clustering] (
     JavaPairRDD.fromRDD(topicDistributions.asInstanceOf[RDD[(java.lang.Long, Vector)]])
   }
 
+  /**
+   * For each document, return the top k weighted topics for that document and their weights.
+   * @return RDD of (doc ID, topic indices, topic weights)
+   */
+  def topTopicsPerDocument(k: Int): RDD[(Long, Array[Int], Array[Double])] = {
+    graph.vertices.filter(LDA.isDocumentVertex).map { case (docID, topicCounts) =>
```
Comment (rotationsymmetry) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: We have `import LDA._` on line 417. So I think we don't need to put `LDA.` for `isDocumentVertex`.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 2]
```diff
@@ -115,6 +115,15 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
       assert(topicDistribution.toArray.sum ~== 1.0 absTol 1e-5)
     }
 
+    val top2TopicsPerDoc = model.topTopicsPerDocument(2).collect()
+    top2TopicsPerDoc.zip(topicDistributions).foreach {
```
Comment (rotationsymmetry) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: The order of documents in `top2TopicPerDoc` and `topicDistribution` are not guaranteed (Please correct me if I am wrong). Recommend we sort them before comparing them.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 3]
```diff
@@ -115,6 +115,15 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
...
+      case ((docId1, indices, weights), (docId2, topicDistribution)) =>
+        assert(indices.length == 2)
+        assert(weights.length == 2)
+        assert(docId1 == docId2)
+        assert(argtopk(topicDistribution.toBreeze, 2).toArray === indices)
+    }
```
Comment (rotationsymmetry) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: Shall we add one more `assert` to check the weights returned by `topTopicsPerDocument`?

**Comment by rotationsymmetry** [comment]
@jkbradley @hhbyyh I just added some comments. Thanks!

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala (hunk 1 above)
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala]: Thanks for noticing it. This is just for code readability and consistency.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala (hunk 2 above)
Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: I've tested it many times and they are of the same order. Anyway, I can rewrite it using join if that looks better.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala (hunk 3 above)
Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: I'll add it. Thanks.

**Comment by rotationsymmetry** [comment]
LGTM.

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala [hunk 4]
```diff
@@ -115,6 +115,17 @@ class LDASuite extends SparkFunSuite with MLlibTestSparkContext {
       assert(topicDistribution.toArray.sum ~== 1.0 absTol 1e-5)
     }
 
+    val top2TopicsPerDoc = model.topTopicsPerDocument(2).map(t => (t._1, (t._2, t._3)))
+    model.topicDistributions.join(top2TopicsPerDoc).collect().foreach {
+      case (docId, (topicDistribution, (indices, weights))) =>
+        assert(indices.length == 2)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/LDASuite.scala]: In the future, please use triple equals in test suites ```===```.  (It generally does not matter, but we're trying to be consistent.)

**Comment by jkbradley** [comment]
Let's just fix that later.  LGTM, merging with master.
Thanks!

**Pull request closed without merging**